File: src/types/contest.ts
import type { PhotoExif } from './gallery';

export type VoteCategory = 'theme' | 'favorite' | 'wildcard';

export interface CategoryVotes {
  theme: number;
  favorite: number;
  wildcard: number;
}

export interface ContestSubmission {
  id: number;
  url: string;
  title: string;
  photographer: string;
  isAssigned?: boolean;
  votes?: number;
  exif?: PhotoExif;
  categoryVotes?: CategoryVotes;
}

export type ContestStatus = 'upcoming' | 'active' | 'voting' | 'completed';

export interface ContestWinner {
  submissionId: number;
  place: 1 | 2 | 3;
  category: VoteCategory;
}

export interface HonorableMention {
  submissionId: number;
  category: VoteCategory;
}

export interface Contest {
  id: number;
  month: string;
  theme: string;
  description: string;
  status: ContestStatus;
  deadline: string;
  submissionCount: number;
  participantCount: number;
  guidelines: string[];
  wildcardCategory?: string | null;
  isImported?: boolean;
  submissions: ContestSubmission[];
  winners?: ContestWinner[];
  honorableMentions?: HonorableMention[];
  userSubmissionCount?: number | null;
  userHasVoted?: boolean | null;
}

export function getCategoryLabel(cat: VoteCategory, wildcardLabel?: string | null): string {
  switch (cat) {
    case 'theme':
      return 'Best Addresses the Theme';
    case 'favorite':
      return 'Personal Favorite';
    case 'wildcard':
      return wildcardLabel || 'Bonus Challenge';
  }
}

// --- My Results types ---

export interface SubmissionResult {
  submissionId: number;
  url: string;
  title: string;
  photographer: string;
  place?: number | null;
  exif?: PhotoExif | null;
}

export interface CategoryResult {
  hasSubmission: boolean;
  bestPlace?: number | null;
  submissions: SubmissionResult[];
}

export interface MyResultsContest {
  contestId: number;
  month: string;
  theme: string;
  wildcardCategory?: string | null;
  hasWildcard: boolean;
  themeResult: CategoryResult;
  favoriteResult: CategoryResult;
  wildcardResult: CategoryResult;
}

export interface LeaderboardRanking {
  value: number;
  rank: number;
  totalMembers: number;
}

export interface MyResultsStats {
  totalSubmissions: number;
  totalVotes: number;
  firstPlaceFinishes: number;
  secondPlaceFinishes: number;
  thirdPlaceFinishes: number;
  podiumFinishes: number;
  contestsEntered: number;
  totalCompletedContests: number;
  participationRate: number;
  bestCategory?: string | null;
}

export interface MyResultsLeaderboard {
  firstPlace: LeaderboardRanking;
  secondPlace: LeaderboardRanking;
  thirdPlace: LeaderboardRanking;
  totalPodium: LeaderboardRanking;
  totalVotes: LeaderboardRanking;
}

export interface MyResultsResponse {
  stats: MyResultsStats;
  leaderboard: MyResultsLeaderboard;
  contests: MyResultsContest[];
}

File: backend/app/schemas/contest.py
from .common import CamelModel


class SubmissionExifSchema(CamelModel):
    camera: str | None = None
    focal_length: str | None = None
    aperture: str | None = None
    shutter_speed: str | None = None
    iso: int | None = None


class CategoryVotesSchema(CamelModel):
    theme: int = 0
    favorite: int = 0
    wildcard: int = 0


class ContestSubmissionResponse(CamelModel):
    id: int
    url: str
    title: str
    photographer: str
    is_assigned: bool = False
    is_own: bool = False
    votes: int | None = None
    exif: SubmissionExifSchema | None = None
    category_votes: CategoryVotesSchema | None = None


class ContestWinnerSchema(CamelModel):
    submission_id: int
    place: int
    category: str = "theme"


class HonorableMentionSchema(CamelModel):
    submission_id: int
    category: str = "theme"


class ContestResponse(CamelModel):
    id: int
    month: str
    theme: str
    description: str
    status: str
    deadline: str
    submission_count: int
    participant_count: int
    guidelines: list[str]
    wildcard_category: str | None = None
    is_imported: bool = False
    submissions: list[ContestSubmissionResponse]
    winners: list[ContestWinnerSchema] | None = None
    honorable_mentions: list[HonorableMentionSchema] | None = None
    user_submission_count: int | None = None
    user_has_voted: bool | None = None


class ContestCreate(CamelModel):
    month: str
    theme: str
    description: str
    status: str = "active"
    deadline: str
    guidelines: list[str]
    wildcard_category: str | None = None


class ContestUpdate(CamelModel):
    month: str | None = None
    theme: str | None = None
    description: str | None = None
    status: str | None = None
    deadline: str | None = None
    guidelines: list[str] | None = None
    wildcard_category: str | None = None


class CategoryVoteRequest(CamelModel):
    category: str
    submission_ids: list[int]


class BatchVoteRequest(CamelModel):
    votes: list[CategoryVoteRequest]


# --- Admin import schemas ---


class SubmissionVoteTally(CamelModel):
    submission_id: int
    theme: int = 0
    favorite: int = 0
    wildcard: int = 0


class FinalizeContestRequest(CamelModel):
    vote_tallies: list[SubmissionVoteTally]


class SubmissionAssignRequest(CamelModel):
    member_id: int | None = None
    photographer: str


# --- My Results schemas ---


class SubmissionResultSchema(CamelModel):
    """A single submission's result within a category."""

    submission_id: int
    url: str
    title: str
    photographer: str
    place: int | None = None
    exif: SubmissionExifSchema | None = None


class CategoryResultSchema(CamelModel):
    """All of the user's submissions for one (contest, category) cell."""

    has_submission: bool
    best_place: int | None = None
    submissions: list[SubmissionResultSchema] = []


class MyResultsContestSchema(CamelModel):
    contest_id: int
    month: str
    theme: str
    wildcard_category: str | None = None
    has_wildcard: bool
    theme_result: CategoryResultSchema
    favorite_result: CategoryResultSchema
    wildcard_result: CategoryResultSchema


class LeaderboardRankingSchema(CamelModel):
    value: int
    rank: int
    total_members: int


class MyResultsStatsSchema(CamelModel):
    total_submissions: int
    total_votes: int
    first_place_finishes: int
    second_place_finishes: int
    third_place_finishes: int
    podium_finishes: int
    contests_entered: int
    total_completed_contests: int
    participation_rate: float
    best_category: str | None = None


class MyResultsLeaderboardSchema(CamelModel):
    first_place: LeaderboardRankingSchema
    second_place: LeaderboardRankingSchema
    third_place: LeaderboardRankingSchema
    total_podium: LeaderboardRankingSchema
    total_votes: LeaderboardRankingSchema


class MyResultsResponseSchema(CamelModel):
    stats: MyResultsStatsSchema
    leaderboard: MyResultsLeaderboardSchema
    contests: list[MyResultsContestSchema]
